### ISSUE TITLE
Policy to enforce MFA when logging in with a 3rd party (federation)

### ISSUE THREAD

**Issue by voidmain** [FusionAuth/fusionauth-issues#2005]
### Problem
Currently, if a user logs in with a 3rd party using federated authentication (Google, Facebook, OIDC, SAML, etc), FusionAuth skips MFA. This assumes that the 3rd party is providing the necessary MFA for the user to ensure they are who they say they are. It's possible that the user has not enabled MFA with the 3rd party but has enabled MFA with FusionAuth.

### Solution
It would be nice to provide a policy for an Identity Provider or a single factor of a user that requires the user to MFA with FusionAuth, regardless of the 3rd party authentication.

### Community guidelines
All issues filed in this repository must abide by the [FusionAuth community guidelines](https://fusionauth.io/community/forum/topic/1000/code-of-conduct).

### How to vote
Please give us a thumbs up or thumbs down as a reaction to help us prioritize this feature. Feel free to comment if you have a particular need or comment on how this feature should work.

**Comment by Thewhiteevil** [comment]
When users enabled authenticator otp, if they try to login through sso, sso must need to ask authenticator otp before redirecting to the dashboard.

**Reference:**

You can see twitter asking for Authenticator OTP while using SSO (not only twitter there are many companies fixing it.)

https://drive.google.com/file/d/1kZjw0unO9jFGPmGGZwhVjYxyEvHdrmvE/view?usp=sharing

**Cheers!**

**Comment by ATLJLawrie** [comment]
Should I open a new request if we would like to see a feature in which logging in with a 3rd party negated the subsequent prompt noted in [ Federated users in an SSO session are prompted to setup MFA on subsequent logins #2357 ](https://github.com/FusionAuth/fusionauth-issues/issues/2357) i.e. an additional tenant / application setting of "Required. A challenge will be required during login _unless a user succesfully authentications via a third-party Identity Provider_" or could this one be modified to include that?

**Comment by Thewhiteevil** [comment]
Yes, sure. Why not? But help me when help is needed. Thank you so much!

**Comment by Thewhiteevil** [comment]
Oh, Nice to see,  you people came to know about the fusion auth bug, which
was submitted by me. Looking for a solution soon.

**Comment by Thewhiteevil** [comment]
Yes, please help me. Thank you so much!

**Comment by mooreds** [comment]
@ATLJLawrie I think we can accommodate your suggestion in this issue.

**Comment by Thewhiteevil** [comment]
Hello team, this is a valid issue to be fixed and Calcom has already fixed it. I plan
to contribute a CVE for this issue soon. Are there any plans on your side
to fix it? Here’s the pull request link:
https://github.com/calcom/cal.com/pull/7509.

**Comment by mooreds** [comment]
@Thewhiteevil this is not on our current roadmap. 

We clearly document the interaction between federated login and MFA: https://fusionauth.io/docs/lifecycle/authenticate-users/identity-providers/#account-security

Appreciate your feedback and we'll take it into account as we look at future product features.

**Comment by mooreds** [comment]
This is now possible using the [MFA requirement lambda](https://fusionauth.io/docs/extend/code/lambdas/mfa-requirement).

```
function checkRequired(result, user, registration, context) {
  var federatedAuthTypes = [
    'APPLE', 'EpicGames', 'FACEBOOK', 'GOOGLE', 'HYPR',
    'LINKEDIN', 'Nintendo', 'OPENID_CONNECT', 'SAMLv2',
    'SAMLv2IdpInitiated', 'SonyPSN', 'Steam', 'TWITTER',
    'Twitch', 'Xbox'
  ];

  if (federatedAuthTypes.indexOf(context.authenticationType) !== -1) {
    result.required = true; // force MFA for federated/social/external IdP logins
  }
}
```

We don't plan to create a policy for this, since the lambda gives you flexibility. 

Closing as not planned.
